### PR TITLE
Move hardtanh activation to Aten(CPU, CUDA)

### PR DESCRIPTION
**Pull request by XiaobingSuper**
@VitalyFedyunin, This PR is about port Hardtanh activation to Aten:
**Test script:**
```
import torch
import torch.nn as nn
import time

torch.manual_seed(0)
def _time():
    if torch.cuda.is_available():
        torch.cuda.synchronize()
    return time.time()

device = "cpu"
m = nn.Hardtanh()
if torch.cuda.is_available():
    device = "cuda"
    m = m.cuda()

#warm up
for n in [100, 10000]:
    input = torch.randn(128, n, requires_grad=True, device=device)
    grad_output = torch.ones(128, n, device=device)
    for i in range(1000):
        output = m(input)
        output.backward(grad_output)

for n in [100, 10000]:
    input = torch.randn(128, n, requires_grad=True, device=device)
    grad_output = torch.ones(128, n, device=device)
    fwd_t = 0
    bwd_t = 0
    for i in range(10000):
        t1 = _time()
        output = m(input)
        t2 = _time()
        output.backward(grad_output)
        t3 = _time()
        fwd_t = fwd_t + (t2 -t1)
        bwd_t = bwd_t + (t3 - t2)
    fwd_avg = fwd_t / 10000 * 1000
    bwd_avg = bwd_t / 10000 * 1000
    print("input size(128, %d) forward time is %.2f (ms); backwad avg time is %.2f (ms)."
          % (n, fwd_avg, bwd_avg))
```
Test Device: CPU: skx-8180, GPU: Tesla P40.
Perfromance:
Before:
```
GPU:
input size(128, 100) forward time is 0.05 (ms); backwad avg time is 0.11 (ms).
input size(128, 10000) forward time is 0.06 (ms); backwad avg time is 0.17 (ms).
CPU
input size(128, 100) forward time is 0.02 (ms); backwad avg time is 0.06 (ms).
input size(128, 10000) forward time is 0.84 (ms); backwad avg time is 0.44 (ms).
```
After:
```
GPU:
input size(128, 100) forward time is 0.05 (ms); backwad avg time is 0.11 (ms).
input size(128, 10000) forward time is 0.06 (ms); backwad avg time is 0.17 (ms).
CPU
input size(128, 100) forward time is 0.02 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 0.61 (ms); backwad avg time is 0.10 (ms).
```
`OMP_NUM_THREADS=1:`
```
Before:
input size(128, 100) forward time is 0.05 (ms); backwad avg time is 0.07 (ms).
input size(128, 10000) forward time is 5.21 (ms); backwad avg time is 5.25 (ms).
After:
input size(128, 100) forward time is 0.01 (ms); backwad avg time is 0.02 (ms).
input size(128, 10000) forward time is 1.09 (ms); backwad avg time is 1.09 (ms).
```

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
#24574 Migrate hardtanh and hardtanh_ from the TH to Aten (CUDA)
#24575 Migrate hardtanh_backward from the TH to Aten (CUDA)
#24710 Migrate hardtanh and hardtanh_ from the TH to Aten (CPU)
#24711 Migrate hardtanh_backward from the TH to Aten (CPU)

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 1]
```diff
@@ -227,6 +227,20 @@ void hardshrink_backward_cpu_kernel(TensorIterator& iter, Scalar lambd) {
...
+void hardtanh_backward_kernel(TensorIterator& iter, Scalar min, Scalar max) {
+  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "hardshrink_backward_cpu", [&] {
+    auto min_val = min.to<scalar_t>();
+    auto max_val = max.to<scalar_t>();
+    cpu_kernel_vec(
+        iter,
+        [=](scalar_t grad_val, scalar_t self_val) {
+          return (self_val <= min_val || self_val >= max_val) ? scalar_t(0) : grad_val;
+        },
+        [=](Vec256<scalar_t> grad_val, Vec256<scalar_t> self_val) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/Activation.cpp]: Please add correctness test which covers backward pass for large tensors.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Overall good, but would be nice to increase test coverage.

**Comment by VitalyFedyunin** [comment]
Also it is unfair to compare single threaded AT code with multi-threaded TH. Can you please add one more section in BEFORE: benchmarks.

File: aten/src/ATen/native/cpu/Activation.cpp (hunk 1 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/cpu/Activation.cpp]: Thanks, I will add test case.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, I add one backward test case for large tensor, please help see it.

**Comment by XiaobingSuper** [comment]
> Also it is unfair to compare single threaded AT code with multi-threaded TH. Can you please add one more section in BEFORE: benchmarks.

I am confused it, I setting the  **OMP_NUM_THREADS** by following script:
```
num_threads=$1
script=$2
last_core=`expr $num_threads - 1`
echo "using $num_threads OMP threads"
echo "bind cores to 0~$last_core"
export OMP_NUM_THREADS=$num_threads
export KMP_AFFINITY=granularity=fine,compact,1,0
numactl --physcpubind=0-$last_core --membind=0 python $script
```
if it is wrong, can you give me some advice how to compare the single thread perfromace? Thanks!

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**